### DISCUSSION TITLE
Canonical identifying publicKey should be Ed25519

### DISCUSSION THREAD

**Discussion by Nuhvi** [renostr/nrps#1] [Ideas]
Using the Ed25519 key as the public key a user is known as, brings the most compatibility possible with existing tools, and overlay networks, too many to enumerate, but for example, the biggest DHT (Mainline DHT used in BitTorrent clients) uses Ed25519 exclusively and would offer a very robust service providers discovery [see this key published on Mainline, stating my Matrix and Nostr handles](https://pkarr.nuh.dev/?pk=o4dksfbqk85ogzdb5osziw6befigbuxmuxkuxq8434q89uj56uyy) 

That `root key` is only needed to sign attestations to other `signing keys`, so no matter how slow it is in comparison to other primitives (it it that slower), that shouldn't factor in the choice, as long as it is secure enough to be used as a long-lasting identifier.

**Comment by dr-orlovsky** [comment]
Agree, just discussed that in Twitter: https://twitter.com/dr_orlovsky/status/1656652626858217477?s=61&t=9A8uvggqKVKV3sT4HPlQyg

**Reply by dr-orlovsky** [reply]
Oh, it was actually you with whom I was talking on Twitter :)

**Reply by Nuhvi** [reply]
Yes, thought it is better to have it here away from the Twitter noise

**Comment by nostronaut** [comment]
a counter-argument, using non-bitcoin EC makes it hard to use adaptor signatures between bitcoin transactions and nostr events

**Reply by Nuhvi** [reply]
I would hope users don't use the same keys to sign bitcoin transactions AND social media notes. Identity keypair can just sign an attestation to many other keys, each for a specific usecase, and even rotate them when you are not comfortable with the security of any key

**Reply by nostronaut** [reply]
That is not what we're talking about. Adaptor signatures as a one-time veriafiable encrypted signature used for atomicity between a pair of transactions. The adaptor being the one time encryption key, if you wish:

https://bitcoinops.org/en/topics/adaptor-signatures/

**Reply by Nuhvi** [reply]
Ok, again, who said you can't use secp256k1 to sign events, do so if you want, but you don't _have_ to use the same key to sign events and control the root of the identity.

The root of the identity being ed25519, means you can lookup records, relays, and providers corresponding to that identity, from any overlay network that supports ed25519 keys, which happens to be almost every single one of them, the same can't be said about any other curve.

Ed25519 for identity, anything else for anything else. Would that satisfy your use case? @nostronaut

**Reply by melvincarvalho** [reply]
After 2.5 years experience of nostr, I also questioned the choice of key and signature at the start.  Actually I think it turned out to be one of the most fortunate decisions that was made, in retrospect.

Using taproot and schnorr aligns with a big community adjacent using open source money.  The jump from bitcoin to nostr and back again was psychologically a low bar, leading to a thriving developer and user community.  

Using bitcoin keys doesnt prevent having other keys, and also registering them in your metadata profile.  I'm working on [noskey](https://melvincarvalho.github.io/noskey/docs/) that will let you have both keys.  We should add a field in type = events that lets you put in an ed25519 key.  Perhaps something as simple has user.ed25519pubkey, assuming there's a common pubkey format

Relays will make it as easy to fetch that key as to get a name, profile etc.  and it's already signed.  I dont necessarily buy the argument that the signature is so hard to verify that you end up trusting the relay, and the relay will go rouge.

**Reply by melvincarvalho** [reply]
What I think I will do, is give every nostr user a derived ED key, which is signed and put in the profile, I think with the field `ed25519pubkey`

Then I'll add a field on https://nostr.social to let users update their profile.  This will let them use things like Git, SSH, other systems, that dont yet support bitcoin, as opt-in.  That keeps a good laser focused eco system together, with small branches out into new use cases.  Then, if any one use case goes viral, it goes viral for all of nostr.  

list of current metadata fields from my dataset:

```
following, 
relay, 
name, 
about, 
picture, 
lud06, 
lud16, 
website, 
display_name, 
nip05, 
banner, 
username, 
badges, 
displayName, 
reactions, 
nip05valid, 
identities, 
webOfTrustScore
```

**Reply by Nuhvi** [reply]
@melvincarvalho I hope you are aware that you didn't engage with the point of the original post. Again, the identifier not being ed25519, make any URL incompatible with existing p2p overlay network, which makes it 10x harder to use as a URL resolvable through a censorship resistance network.

You didn't engage with that or explain why should the root identity be secp and point to Ed keys instead of the other way around. I assume it is just bias and hope that existing nost public keys stay the long lasting identities. 

I am OK with that, if you offer a solution to discovery that is better and more proven than pkarr.nuh.dev

**Reply by Nuhvi** [reply]
My general conclusion really is that Nostr devs don't really share my opinion that we need public keys to be resolvable through something less silly than a hard-coded list of relays, and if that is true, then nothing I say applies really.

**Reply by Nuhvi** [reply]
However if you share my opinion that it is worth having public keys as URLs that resolves to hosting providers and other sub keys, then please show me anything that can do that better and more proven than Bittorrent.

Here is my `ed25519` key pointing to my Matrix ID and npub:
https://pkarr.nuh.dev/?pk=o4dksfbqk85ogzdb5osziw6befigbuxmuxkuxq8434q89uj56uyy

What else is missing here exactly? Because I know you can't do that with a Nostr key, without either hardcoding a relay in the URL (destroying the purpose of urls), or hoping that the profile you are looking for, was posted to one of the relays you follow! There is a reason the Web doesn't work like that, because that is silly.

**Reply by melvincarvalho** [reply]
Clients are not struggling to resolve public keys through the relay infrastructure.  And REST caching infrastructural is being added to that.  There could be other resolution functions because nostr IDs are truly decentralized.  If there's a need, it will be met.  Presently nobody has reported any issues on that front.  Im glad folks are thinking about it, but it's an edge case if anything right now.

**Reply by melvincarvalho** [reply]
> However if you share my opinion that it is worth having public keys as URLs that resolves to hosting providers and other sub keys, then please show me anything that can do that better and more proven than Bittorrent.
> 
> Here is my `ed25519` key pointing to my Matrix ID and npub: https://pkarr.nuh.dev/?pk=o4dksfbqk85ogzdb5osziw6befigbuxmuxkuxq8434q89uj56uyy
> 
> What else is missing here exactly? Because I know you can't do that with a Nostr key, without either hardcoding a relay in the URL (destroying the purpose of urls), or hoping that the profile you are looking for, was posted to one of the relays you follow! There is a reason the Web doesn't work like that, because that is silly.

This is horrific NIH.  Look at mine:

- nostr key: https://nostr.social/de7ecd1e2976a6adb2ffa5f4db81a7d812c8bb6698aa00dcf1e76adb55efd645
- linked data: https://nostr.social/.well-known/nostr/pubkey/de7ecd1e2976a6adb2ffa5f4db81a7d812c8bb6698aa00dcf1e76adb55efd645/index.json
- vocab : https://w3id.org/nostr
- context : https://w3id.org/nostr/context
- nip05: https://melvincarvalho.com/.well-known/nostr.json
- homepage: https://melvincarvalho.com/
- includes matrix ID
- includes friends IDs, social graph
- includes apps
- includes on chain commitments
- points to git commitments
- includes single use seals: http://gitmark.info/c06e4b12945a0c279900c9d80491177d45fcf2971893a60a19afcf5c9604fb16

Nostr IDs are very rich for discovery.

**Reply by Nuhvi** [reply]
Good

**Reply by dr-orlovsky** [reply]
> After 2.5 years experience of nostr, I also questioned the choice of key and signature at the start. Actually I think it turned out to be one of the most fortunate decisions that was made, in retrospect.

Can't agree with you, @melvincarvalho, here. Use of Schnorr signatures leaks private key on key re-use. While in Bitcoin you can avoid key reuse (and it is recommended to do so to improve privacy), the situation with nostr is the opposite: you can't avoid re-using the same key thousands - and even hundreds of thousand times, meaning that eventually each popular Nostr account will be hacked in a matter of year or so.

**Reply by melvincarvalho** [reply]
@dr-orlovsky hmm, good point.  How big is the nonce used, and how many uses are needed to create a collision?  1 year seems low, but maybe you are right.  Needs verification.

**Reply by melvincarvalho** [reply]
Some quick research:

> a 256-bit nonce is typical (as used in Bitcoin's implementation of the Schnorr signature)

If that is correct, and then pure nonce collisions are unlikely to happen in one year

There might be some other attacks from key reuse though.  If keys last 100+ years, or 10+ years there will be a chance to develop rotation.  If it's <1 year, that might be more of a problem.  But no key has been cracked yet.

**Reply by dr-orlovsky** [reply]
It is not about the nonce collisions but about a bias in the nonce randomness 1-2 bits of bias (in all 256 bits, i.e. < 1% of bias) is enough to leak the private key over many thousands of sigs (in nostr opinion leaders may have millions of them).

1-2 bit bias - is much lower than JavaScript-based pseudorandom generations, which means most of nostr clients are vulnerable.

**Reply by melvincarvalho** [reply]
@dr-orlovsky do you have a link that explains this vulnerability?

**Reply by dr-orlovsky** [reply]
https://ecc2017.cs.ru.nl/slides/ecc2017-tibouchi.pdf
